Event tracking fixes (#4672)
bbc/simorgh: src/app/containers/ATIAnalytics/README.md

This directory contains code for logging events to ATI Analytics.

## Logging Click Events
For logging click events you will need four things:

> * [useClickTracker](https://github.com/bbc/simorgh/blob/latest/src/app/contexts/EventContext/index.jsx) used to track a component to be tracked. This should be done by using `data-attributes` used in the component. This data attribute should also contain value for building the component info object, e.g. `data-navigation=‘Akuko_0’`

Reviewer: ```suggestion
* [useClickTracker](https://github.com/bbc/simorgh/blob/latest/src/app/contexts/EventContext/index.jsx) used to add click event to the component to be tracked. This should be done by using `data-attributes` used in the component. This data attribute should also contain value for building the component info object, e.g. `data-navigation=‘Akuko_0’`
```

* to compose the tracking URL via [buildATIClickParams](https://github.com/bbc/simorgh/blob/latest/src/app/containers/ATIAnalytics/params/index.js#L41) which uses [RequestContext](https://github.com/bbc/simorgh/blob/latest/src/app/contexts/RequestContext/index.jsx) and [ServiceContext](https://github.com/bbc/simorgh/blob/latest/src/app/contexts/ServiceContext/index.jsx).
* to get all the component information by using [getComponentInfo](https://github.com/bbc/simorgh/blob/latest/src/app/lib/analyticsUtils/index.js)
* [sendEventBeacon](https://github.com/bbc/simorgh/blob/latest/src/app/containers/ATIAnalytics/beacon/index.js) for sending the two events - ‘click’ and ‘view’.

## How to set up tracking for a component
In the container for the component you’re trying to track,  import `EventContext`, and use `useClickTracker` as such:

```

> import React, { useContext } from ‘react’;

Reviewer 2: There are a few places in this file where there's a ‘ rather than a '. Text editors 🙄
```suggestion
import React, { useContext } from 'react';
```

import { EventContext } from ‘#contexts/EventContext’;

const componentContainer = () => {
	const { useClickTracker } = useContext(EventContext);

	useClickTracker('[data-attr]’, event => {
		// `sendEventBeacon` will go here
	});
}
```

Provide `useClickTracker` with the data-attribute attached to the component to be tracked, e.g. for Navigation component that would be `data-navigation`. This data attribute should be set in the component's container and passed down to the Psammead component.

If not done already, Request and Service Contexts will need to be imported to create the tracking URL. To do so, import `buildATIClickParams` and use it within the container to get props required for sending the event beacon.

> ```

Reviewer: ```suggestion
```jsx
```
Enables some syntax highlighting goodness. 😄

	const eventTrackingProps = buildATIClickParams(
		{},
		useContext(RequestContext),
    	useContext(ServiceContext)
	);

	useClickTracker('[data-navigation]’, event => {
		const eventData = event.srcElement.dataset.navigation;
		const componentDataSplit = eventData.split(‘_’);
		const componentData = {
			creationLabel: componentDataSplit[0],
	      child: componentDataSplit[1],
	   };
	
		const componentInfo = getComponentInfo(event, ‘navigation’, componentData);

		// `sendEventBeacon` will go here
	});
}
```

All this information can then be used to put together the event beacon.

```
import React, { useContext } from ‘react’;
import { buildATIClickParams } from ‘../ATIAnalytics/params’;
import { EventContext } from ‘#contexts/EventContext’;
import { RequestContext } from ‘#contexts/RequestContext’;
import { ServiceContext } from ‘#contexts/ServiceContext’;
import { sendEventBeacon } from ‘../ATIAnalytics/beacon’;

const componentContainer = () => {
	const { useClickTracker } = useContext(EventContext);
	const eventTrackingProps = buildATIClickParams(
		{},
		useContext(RequestContext),
    	useContext(ServiceContext)
	);

	useClickTracker('[data-navigation]’, event => {

> 		const eventData = event.srcElement.dataset.navigation;

Reviewer 3: Could we pass the data we need as separate attributes rather than creating and then splitting this string?

Author: Good point. We could do that. I assumed that we would then have to extend the functionality of `useClickTracker` to track more data attrs, but we can get that from the `event`. We can then set data attributes per information that needs to be send

Reviewer 2: I've done a little bit of playing and we can do something along the lines of
```jsx
  useClickTracker('[data-navigation]', event => {
    const componentName = 'navigation';
    const creationLabel = event.srcElement.dataset['creationlabel'];
    const child = event.srcElement.dataset['child'];
    const componentData = {
      creationLabel,
      child,
    };

    const componentInfo = getComponentInfo({
      url: event.target.href,
      componentName,
      componentData,
    });

    sendEventBeacon({
      componentName,
      type: 'click',
      componentInfo,
      ...eventTrackingProps,
    });
  });
```

Of course, this isn't actually in scope of this ticket, but it would be good to figure this out if it's going into the Readme. We could also just put a comment in this bit of the readme saying "// get the relevant data" or something similar until we figure out our preferred approach.

Reviewer 2: I've put this version in the readme but hopefully it's clear from the readme that this is only one way of doing it.

		const componentDataSplit = eventData.split(‘_’);
		const componentData = {
			creationLabel: componentDataSplit[0],
		  child: componentDataSplit[1],
		};
		
		const componentInfo = getComponentInfo(event, ‘navigation’, componentData);

		sendEventBeacon({
	      element: event.target,

> 	     ‘navigation,

Reviewer 2: This one is also missing the closing `'`
```suggestion
	     'navigation',
```

This directory contains code for logging events to ATI Analytics.

## Logging Click Events
For logging click events you will need four things:

> * [useClickTracker](https://github.com/bbc/simorgh/blob/latest/src/app/contexts/EventContext/index.jsx) used to add click event to the to be tracked. This should be done by using `data-attributes` used in the component. This event handler also needs to get the component data ready for the `getComponentInfo` function, which can be done using other data attributes.

Reviewer 4: ```suggestion
* [useClickTracker](https://github.com/bbc/simorgh/blob/latest/src/app/contexts/EventContext/index.jsx) used to add click event to the component to be tracked. This should be done by using `data-attributes` used in the component. This event handler also needs to get the component data ready for the `getComponentInfo` function, which can be done using other data attributes.
```

import { ServiceContext } from '#contexts/ServiceContext';

const componentContainer = (navigationItems) => {
	const { useClickTracker } = useContext(EventContext);
	const eventTrackingProps = buildATIClickParams(
		{},
		useContext(RequestContext),
		useContext(ServiceContext)
	);

	useClickTracker('[data-navigation]', event => {
		// `sendEventBeacon` will go here
	});

	return (
		<div className="navigation">
			<ul className="navigation_list">
				{navigationItems.map(item => (
					<li
						key={item.title}
						data-navigation
					>
						{item.title}
					</li>
				))}
			</ul>
		</div>
	);
}
```

To get all the information about the component and the event, `getComponentInfo` will have to be used. This function takes in 3 parameters - the result, name of the component, and component data.
Component data should be an object containing information obtained from data attributes.
The result is a string of the result of having clicked, eg the url of the link or the name of the video being played.

The component data object should contain a value for `actionLabel`, `child`, and optionally `source`. These values will be used as part of the tracking URL.

> - source: The source of the event at a high level. Examples include an urn, or 'responsive_web'

Reviewer 4: Can we put the `source` in the example below?

Reviewer 2: This is set by default to `responsive_web~news-simorgh`. For setting to an urn, I think that is more for clicking on a video or maybe something in an article, so in the example I think we'd want to leave it as a default.

I've added a comment both to this line and in the example code to try to make that clear.

- child: The specific child of the data that is doing the action - eg button::1
- actionLabel: The name of the action - eg navigation-home

In the navigation example, these values were obtained by using multiple data attributes on the navigation's `li` item.

```jsx
import React, { useContext } from 'react';
import { buildATIClickParams } from '../ATIAnalytics/params';
import { EventContext } from '#contexts/EventContext';
import { RequestContext } from '#contexts/RequestContext';
import { ServiceContext } from '#contexts/ServiceContext';

const componentContainer = () => {
	const { useClickTracker } = useContext(EventContext);
	const eventTrackingProps = buildATIClickParams(
		{},
		useContext(RequestContext),
		useContext(ServiceContext)
	);

	  useClickTracker('[data-navigation]', event => {
		const componentName = 'navigation';
		const componentData = {

> 			containerLabel: event.srcElement.dataset.containerlabel,

Reviewer 4: Should this be `actionLabel`?

Reviewer 2: Yep, good spot! Have updated all of these in the readme.

		});

		sendEventBeacon({
			componentName,
			type: 'click',
			componentInfo,
			...eventTrackingProps,
		});

		return (
			<div className="navigation">
				<ul className="navigation_list">
					{navigationItems.map((item, index) => (
						<li
							key={item.title}
							data-navigation
							data-containerlabel={`navigation-${item.title}`}
							data-child={index}
						>
							{item.title}
						</li>
					))}
				</ul>
			</div>
		);
	});
}
```

Now clicking on the nav links should register as a click event in the ATI Tag Inspector.

## Tag Inspector

This ATI system was set up for adverts but repurposed by the BBC for click tracking. Because of this, the names of the fields that are seen in the ATI Tag Inspector are not indicative of the BBC's use of those fields. Below is a table of the ATI label and the corresponding BBC label. In the Simorgh code, we only refer to the BBC label so that the only time discrepancy is seen is in the ATI system. The one exception is with creation/attribute, which has been referred to as `action` in code as this is more representative of the meaning.

| ATi label | BBC label | Usage notes |
| --------- | --------- | ----------- |

> | camapaignId | container | |

Reviewer 4: ```suggestion
| campaignId | container | |
```
